### PR TITLE
realtek: add support for RTL9300/RTL9310 I2C controllers

### PR DESCRIPTION
**Pull request by ghost**
This PR provides drivers support for RTL9300 and RTL9300 I2C controllers.They provide only SMBus compatible I2C transfers and cannot directly manipulate the I2C bus. The controller also provides a multiplexer between two different independent masters each capable of controlling 8 (RTL9300) or 16 busses (RTL9310). An important limitation is that the SCL pin of each master is fixed.
The following is an example .dts for the RTL9310 Edgecore 4125-10P to support the SFP cages:

```
        i2c0: i2c-rtl9310 {
		compatible = "realtek,rtl9310-i2c";
		reg = <0x1b001004 0x3c>;
		#address-cells = <1>;
		#size-cells = <0>;
		clock-frequency = <100000>;
	};

	i2cmux {
		compatible = "realtek,i2c-mux-rtl9310";
		#address-cells = <1>;
		#size-cells = <0>;
		i2c-parent = <&i2c0>;

		i2c01: i2c-rtl9310-1 {
			reg = <1>;
			#address-cells = <1>;
			#size-cells = <0>;
			sda-pin = <4>;
			scl-pin = <13>;
		};

		i2c02: i2c-rtl9310-2 {
			reg = <2>;
			#address-cells = <1>;
			#size-cells = <0>;
			sda-pin = <5>;
			scl-pin = <13>;
		};
	};

```
At present it is not understood which physical pins actually are used for SDA on the RTL9310. The driver merely makes the pin-ids configurable, which have an unknown mapping to physical pins. However, these pin-ids are also what is provided as information by the OEM firmware so that any device can be configured. This also prevents usage of pinmuxes, which might be more elegant and could also provide support for OpenDrain bus termination.
Another limitation is that a standard linux multiplexer lock is being used, so that the 2 masters cannot do transfers in parallel.
There has not been any testing at the physical level of the bus, however the driver allows to successfully access the SFP EEPROMS on the RTL9300-based Ubiquiti USW and the Edgecore ECS4125-10P (RTL9310). On the Ubiquiti also PCA9555a GPIO expanders can be steered successfully, on the Edgecore an LM thermal sensor read.

### CODE REVIEW

File: target/linux/realtek/config-5.10 [hunk 1]
```diff
@@ -85,7 +85,9 @@ CONFIG_HZ_PERIODIC=y
 CONFIG_I2C=y
 CONFIG_I2C_ALGOBIT=y
 CONFIG_I2C_BOARDINFO=y
+CONFIG_I2C_CHARDEV=y
 CONFIG_I2C_GPIO=y
+CONFIG_I2C_RTL9300=y
```
Comment (svanheule) [inline target/linux/realtek/config-5.10:92]: This enables both the driver in this commit, as well as userspace access to I2C devices. In my opinion, at least the latter should be a separate commit with argumentation why it needs to be enabled.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c [hunk 2]
```diff
@@ -0,0 +1,484 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <linux/module.h>
+#include <linux/of_platform.h>
+#include "i2c-rtl9300.h"
+
+#define REG(x)		(i2c->base + x + (i2c->scl_num ? 0x1c : 0))
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:7]: What is i2c? There's no context as to what it is, here with this macro.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+#define REG_MASK(clear, set, reg)	\
+			writel((readl(REG(reg)) & ~(clear)) | (set), REG(reg))
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:9]: I guess it lowers your line count, but I don't think it makes the code very easy to read, when you use macros like these.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+	void (*config_io)(struct rtl9300_i2c *i2c, int scl_num, int sda_num);
+};
+			
+DEFINE_MUTEX(i2c_lock);
+
+static void rtl9300_i2c_reg_addr_set(struct rtl9300_i2c *i2c, u32 reg, u16 len)
+{
+	// Set register address width
+	REG_MASK(0x3 << RTL9300_I2C_CTRL2_MADDR_WIDTH, len << RTL9300_I2C_CTRL2_MADDR_WIDTH,
+		 RTL9300_I2C_CTRL2);
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:32]: I really prefer a few #define-s to document what the register fields are, and then using FIELD_PREP, FIELD_GET,... from bitfield.h.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+	u32 v;
+
+	// Set SCL pin
+	REG_MASK(0, BIT(RTL9300_I2C_CTRL1_GPIO8_SCL_SEL), RTL9300_I2C_CTRL1);
+
+	// Set SDA pin
+	REG_MASK(0x7 << RTL9300_I2C_CTRL1_SDA_OUT_SEL,
+		 i2c->sda_num << RTL9300_I2C_CTRL1_SDA_OUT_SEL, RTL9300_I2C_CTRL1);
+
+	// Set SDA pin to I2C functionality
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:60]: This smells like it should be in a pin control driver, rather than an I2C device driver.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+	v = readl(i2c->base + RTL9300_I2C_MST_GLB_CTRL);
+	v |= BIT(i2c->sda_num);
+	writel(v, i2c->base + RTL9300_I2C_MST_GLB_CTRL);
+}
+
+static void rtl9310_i2c_config_io(struct rtl9300_i2c *i2c, int scl_num, int sda_num)
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:66]: This appears to have the same logic as rtl9300_i2c_config_io, just with different offsets. Can't these be stored in in e.g. rtl9300_i2c, so the same logic can be used?

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+
+	if (len > 16)
+		return -EIO;
+
+	for (i = 0; i < len; i++) {
+		if (i % 4 == 0)
+			v = readl(r0 + i);
+		buf[i] = v;
+		v >>= 8;
+	}
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:136]: Why at most 16 ~longs~ bytes? What's wrong with reading bytes (ioread8, readb) instead of rolling your own aligned-reads?

File: target/linux/realtek/config-5.10 (hunk 1 above)
Comment (bkobl) [reply target/linux/realtek/config-5.10:92]: I'll split this up. The /dev/i2c- devices are extremely useful, as they allow to use the i2c tools, e.g. to identify the SFP modules, fans and thermal probes as well as other tools accessing these devices. Up to until now we did not have a i2c device available on the platform, only sfp modules accessed via gpio, which did not expose an i2c device usable outside of the sfp module driver.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 2 above)
Comment (bkobl) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:7]: I'll change this to #define REG(i, x)		(i->base + x + (i->scl_num ? 0x1c : 0))
and elsewhere REG(x) -> REG(i,x)
That was quite minimalistic, I agree. I will also pick the 0x1c from a variable in i2c.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 3 above)
Comment (bkobl) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:9]: This is pretty standard fare in device drivers in the kernel. I have seen it a couple of times, e.g. with a quick grep:
pad_w32_mask() in  drivers/pinctrl/pinctrl-falcon.c
xway_stp_w32_mask in drivers/gpio/gpio-stp-xway.c 
snd_hdac_updatel in include/sound/hdaudio_ext.h
and I think I got it from John, who uses it extensively, e.g. here:
arch/mips/lantiq/falcon/sysctrl.c:#define sysctl_w32_mask(m, clear, set, reg)

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 4 above)
Comment (bkobl) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:32]: RTL9300_I2C_CTRL2_MADDR_WIDTH is anyway defined. FIELD_PREP and FIELD_GET will just increase the line-count considerably here. The line says quite a lot: RTL9300_I2C_CTRL2_MADDR_WIDTH is a 2-bit field starting at RTL9300_I2C_CTRL2_MADDR_WIDTH, which is really not more than you can anyway get. I'll leave this to the taste of a subsystem maintainer should this be submitted.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 5 above)
Comment (bkobl) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:60]: You cannot do that because it needs to be dynamically controlled by the muxer, see: rtl9300_i2c_mux_select(), which sets RTL9300_I2C_CTRL1_GPIO8_SCL_SEL and RTL9300_I2C_CTRL1_SDA_OUT_SEL depending on the multiplex. The gpio pin is only for the time of the transfer actually an I2C pin and if you configure for a different multiplex this changes again.
But of course the original reason was that writing a pin control driver would be a huge overhead just for 2 pins controlled by this I2C driver. Maybe once we understand the SPI driver, there seem to be a couple of interesting SPI devices in the RTL93xx devices.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 6 above)
Comment (bkobl) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:66]: You are right. But this seems to be a matter of taste. I was using stored offsets extensively in the dsa/ethernet drivers. But people like Sebastian go after them and prefer to have duplicate code called via function pointers like here. The argument seems to be that then you don't see the offset values directly in the code, which is important to verify that it works correctly. Forgetting to set one of the offset values in the structures is a very good way of coding a very difficult to find bug. We were using a redefined REG() macro to find all the instances for the RTL931X, otherwise it is not possible if you have hundreds of offsets on 4 different SoC architectures. The good thing about forgetting to set function pointers is the rather dramatic and immediate crash.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 7 above)
Comment (bkobl) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:136]: The 16 bytes is a common limitation for SMBus drivers. You cannot enlarge this, as the HW is waiting for a reply after sending 16 bytes or will send a reply after receiving 16 bytes. So 2 x 16 byte transfers does not give a 32 byte transfer. References to this limit are found e.g. in the SFP driver. The 16 bytes are spread rather strangely across
RTL9300_I2C_MST1_DATA_WORD1_ADDR, RTL9300_I2C_MST1_DATA_WORD2_ADDR, RTL9300_I2C_MST1_DATA_WORD3_ADDR, RTL9300_I2C_MST1_DATA_WORD4_ADDR
Better to make the reader aware of this instead of pretending to have a single field.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 4 above)
Comment (svanheule) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:32]: Line count by itself is a bad metric. Every next person will have to figure out and understand what this macro means. Using FIELD_PREP is not so bad really:

    u32 v;
    v = readl(RTL9300_IC2_CTRL2) & ~RTL9300_IC2_CTRL_MADDR;
    writel(v | FIELD_PREP(RTL9300_I2C_CTRL_MADDR, len), RTL9300_I2C_CTRL2);

Or if you really want to use a macro:

    #define REG_FIELD_WRITE(offset, mask, val) \
        writel((readl(offset) & ~mask) | FIELD_PREP(mask, val), offset)
    REG_FIELD_WRITE(RTL9300_I2C_CTRL2, RTL9300_I2C_CTRL_MADDR, len);

IMHO `REG_MASK` also just isn't a good name for a macro that clears and sets two arbitrary sets of bits.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 5 above)
Comment (svanheule) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:60]: Selecting the bus master for that pin makes no sense when it is configured as a GPIO pin. So selecting between GPIO and I2C functionality is a separate problem, and should be handled elsewhere. It's true that Realtek's register space is completely bonkers, but there are already some other bit(field)s in use that switch between GPIO or alternate function (e.g. the system LED pin).

Comment (svanheule) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:60]: I've already experimented with a pinctrl driver for the switchcore register space, but I don't think I had figured out how to properly manage concurrent access to the registers that have these pin control bits.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 6 above)
Comment (svanheule) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:66]: Copy-pasting code is also a very good way of creating bugs. Again, I think this only makes the code harder to understand when another person is reading it. This style implements a certain 'device op' for RTL9310 and RTL9300. Most of these are rather shot, but the execute_transfer() op is quite a bit more complicated. If both I2C controllers work in (exactly) the same way, there should only be one implementation of that logic in my opinion.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 7 above)
Comment (svanheule) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:136]: How about something like

    u32 data[4];
    for (i = 0; i < len; i += 4)
        data[i/4] = swab32(readl(r0 + i));
    
    u8 *data_bytes = &data[0];
    for (i = 0; i < len; i++)
        *buf = data_bytes[i];

This would make it very obvious that you're reading registers, reordering the bytes, and then writing them out to the destination buffer.

File: target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c (hunk 6 above)
Comment (svanheule) [reply target/linux/realtek/files-5.10/drivers/i2c/busses/i2c-rtl9300.c:66]: I see now that the registers differ between RTL9300 and RTL9310. They just happen to have the same control bits laid out differently over the two control registers :expressionless:
I still think there should be a better way of abstracting this. With a regmap interface, you could define fields for all the control bits, and put these in a private data structure. Then you would only need one function to implement each piece of logic, and could let those refer to the regmap fields.

**Comment by svanheule** [comment]
Unless I'm missing something, a newer version of these changes was merged with #4973, so I'll close this PR.
